### ISSUE TITLE
[Yul Optimizer] Function outliner

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#6713]
In the sol->yul code generation, we end up with quite a bit of code that contains similar or even identical blocks of code. It would be nice to have an optimizer component that e.g. recognizes common suffixes of functions or common internal blocks which can then be moved to a single function - as long as the number of local variables that are modified in that block but defined outside is limited.

**Comment by ekpyron** [comment]
Consider a block like (with external vars ``b`` and ``c``):
```
{
    let d := add(b, b)
    c := d
}
```

We can compute a hash based on:
```
{
  let <internal0> := add(<external0>, <external0>)
  <external1> := <internal0>
}
```
appending the number of external references (i.e. ``2``) to the hash or storing it alongside the hash - let's call that ``BLOCK0HASH``.

Then for
```
{
    let aa := mload(0)
    let bb := mload(1)
    let cc := mload(2)

    {
        let dd := add(bb, bb)
        cc := dd
    }
    mstore(aa, cc)
}
```
(The inner block will also have hash value ``BLOCK0HASH``)
We can compute a hash like
```
{
    let <internal0> := mload(0)
    let <internal1> := mload(1)
    let <internal2> := mload(2)

    IMPORT_BLOCK(BLOCK0HASH, <internal1>, <internal2>)
    mstore(<internal0>, <internal2>)
}
```

For every block the hash can be stored in a hash map to allow lookups.
For blocks with identical hashs we can compute structural equality.

**Comment by ekpyron** [comment]
We can add the "block hash" of a function body to the RoughHeuristic in ``EquivalentFunctionDetector`` as well.

**Comment by ekpyron** [comment]
Hm... it's still a bit more complicated... the external variables in the block have to be split into function parameters and return values - return values have to be the union of all externally referenced variables that are accessed after any occurrance of the block to be outlined... moreover those variables might not only be written to, but they might have been read from, so they might have to be passed as arguments to the outlined function as well - might be worth considering to require SSA form to alleviate that.
Let alone to find a good heuristics about when to outline equivalent blocks or not, but that'd be a second step. Quite some stuff left to do after detecting equivalent blocks in any case :-).

**Comment by chriseth** [comment]
Perhaps as a first step, we can do the full thing, i.e. outlining the block `B[ext1, ext2, ext3]` as 
```
ext1, ext2, ext3 := f(ext1, ext2, ext3)
```
A second step (that is actually needed anyway) can analyze if a function return variable can be removed because it is not used at any of its call sites (we need a similar thing for parameters, but that analysis would be local to the function).

**Comment by leonardoalt** [comment]
Abstract Syntax Merkle Tree? :p

**Comment by ekpyron** [comment]
Another approach to this would be to have a step that just outlines everything that has a sufficient size and has few inputs and outputs (although we'd still have to detect which groups of statements qualify for that), then simplify and hope that we end up having equivalent functions - then we could inline everything back that's still only used once. That might even help with the stack height problem (if a part of a large function has few inputs and outputs, but lots of local intermediate variables, outlining it might be good, even if the outlined function is used only once).

**Comment by ekpyron** [comment]
Note, mainly to myself: my most recent ideas around this would go along the lines of the following:
Consider:
```
{
  let x := mload(a)
  let y := add(calldataload(b), x)
  b := x
  if y { b := mul(b, 2) }
  let z := mul(y, 2)
  sstore(x, z)
}
```
Move all declarations up front:

```
{
  let x let y let z
  x := mload(a)
  y := add(calldataload(b), x)
  b := x
  if y { b := mul(b, 2) }
  z := mul(y, 2)
  sstore(x, z)
}
```
Introduce "basic control flow blocks":

```
{
  let x let y let z
  {
    x := mload(a)
    y := add(calldataload(b), x)
    b := x
  }
  {
    if y { b := mul(b, 2) }
  }
  {
    z := mul(y, 2)
    sstore(x, z)
  }
}
```
Outline those blocks "naively:
```
function f1(x_in, a_in, y_in, b_in) -> x,a,y,b {
    x := x_in   a := a_in    y := y_in    b := b_in
    x := mload(a)
    y := add(calldataload(b), x)
    b := x
}
function f2(y_in, b_in) -> y, b
{
    y := y_in  b := b_in
    if y { b := mul(b, 2) } // maybe the inner block would be outlined in the same step
}
function f3(z_in, y_in, x_in) -> z, y, x {
    z := z_in    y := y_in    x := x_in
    z := mul(y, 2)
    sstore(x, z)
}
{
  let x let y let z
  x, a, y, b := f1(x, a, y, b)
  y, b := f2(y, b)
  z, y, x := f3(z, y, x)
}
```

Next thing would be to remove redundant arguments and maybe figure out what returns are actually needed. Might be better to start in SSA form or something like that.
Maybe some more simplification at this stage. Then equivalent function detection. Then inlining again.
This could potentially solve the precondition modifier case already. But the basic control flow blocks will still potentially be large and it'll be hard to find patterns in them. Also not sure that this is the best structure for common patterns in general - there are potential patterns this will miss by design.
Still not sure about all this in general.

**Comment by chriseth** [comment]
One very important use-case for this is modifiers: We currently generate code of the form
```
function f() {
  modifier_f_1()
}
function modifier_f_1() {
  ...
  f_inner()
}
```
Where `modifier_f_1` is a function specialized for `f`  - i.e. the modifier code is copied for each of its uses.

At least this special use-case could be detected by the optimizer and outlined.

Maybe we can even do something like this on the opcode level?
